src/wp-includes/class-wp-query.php

```diff
@@ -3476,8 +3476,27 @@ public function get_posts() {
 			$this->posts = array_map( 'get_post', $this->posts );
 
 			if ( $q['cache_results'] ) {
-				$post_ids = wp_list_pluck( $this->posts, 'ID' );
-				_prime_post_caches( $post_ids, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+				$post_ids       = wp_list_pluck( $this->posts, 'ID' );
+				$non_cached_ids = _get_non_cached_ids( $post_ids, 'posts' );
+				if ( ! empty( $non_cached_ids ) ) {
+					$fresh_posts = array();
+					foreach ( $this->posts as $post ) {
+						if ( in_array( $post->ID, $non_cached_ids, true ) ) {
+							$fresh_posts[] = $post;
+						}
+					}
+					update_post_cache( $fresh_posts );
+				}
+
+				if ( $q['update_post_meta_cache'] ) {
+					update_postmeta_cache( $post_ids );
+				}
+
+				if ( $q['update_post_term_cache'] ) {
+					$post_types = wp_list_pluck( $this->posts, 'post_type' );
+					$post_types = array_unique( $post_types );
+					update_object_term_cache( $post_ids, $post_types );
+				}
```

Comment:
Code for reviews, I would love to remove this block of code of outside the `if ( $q['cache_results'] ) {` state. There is undocumented that `update_post_meta_cache` and `update_post_term_cache` only work if `cache_results` is set to true.

Reply:
I was investigating the `cache_results` parameter during the 6.1 cycle and noticed it hadn't done anything for a very long time, possibly since `split_the_query` was introduced.

I think `cache_results` affecting the main query only is fine. It's almost certain `get_post()` will be called during the loop at which case the post will be cached.